core/src/main/java/org/frankframework/statistics/MetricsInitializer.java

```diff
@@ -66,86 +75,147 @@ public void afterPropertiesSet() throws Exception {
-		//not used
+
+
+	private List<Tag> getTags(@Nonnull IConfigurationAware frankElement, @Nonnull String name, @Nullable List<Tag> extraTags) {
+		ApplicationContext configuration = frankElement.getApplicationContext();
+		List<Tag> tags = new ArrayList<>(5);
+		Adapter adapter = getAdapter(frankElement);
+		if(adapter != null) {
+			tags.add(Tag.of("adapter", adapter.getAdapter().getName()));
+		}
+		tags.add(Tag.of("configuration", configuration.getId()));
+		tags.add(Tag.of("name", name));
+		tags.add(Tag.of("type", getElementType(frankElement)));
+		if(extraTags != null) {
+			tags.addAll(extraTags);
+		}
+
+		return tags;
 	}
 
-	@Override
-	public void handleScalar(NodeConfig data, String scalarName, Date value) throws SenderException {
-		//not used
+	private Adapter getAdapter(@Nonnull IConfigurationAware frankElement) {
+		if(frankElement instanceof Adapter) {
+			return (Adapter) frankElement;
+		}
+		if(frankElement instanceof HasStatistics) {
+			HasStatistics elm = (HasStatistics) frankElement;
+			if(elm.getAdapter() != null) {
+				return elm.getAdapter();
+			}
+		}
+		return null;
 	}
 
-	@Override
-	public NodeConfig openGroup(NodeConfig parentData, String dimensionName, String type) throws SenderException {
-		String nodeName = parentData.name;
-		List<Tag> tags = new LinkedList<>(parentData.tags);
-		int groupLevel = parentData.groupLevel;
-		if (StringUtils.isNotEmpty(dimensionName)) {
-			tags.add(Tag.of(type, dimensionName));
+	private String getElementType(@Nonnull IConfigurationAware frankElement) {
+		if(frankElement instanceof Receiver) {
+			return "receiver";
+		} else if(frankElement instanceof PipeLine) {
+			return "pipeline";
+		} else if(frankElement instanceof IPipe) {
+			return "pipe";
+		} else if(frankElement instanceof Adapter) {
+			return "adapter";
+		} else if(frankElement instanceof ISender) {
+			return "sender";
+		} else if(frankElement instanceof JobDef) {
+			return "schedule";
 		} else {
-			nodeName=nodeName+"."+type;
+			throw new IllegalStateException("meter type not configured");
 		}
-		return new NodeConfig(nodeName, tags, groupLevel);
 	}
 
-	@Override
-	public void closeGroup(NodeConfig data) throws SenderException {
-		//not used
+	private double[] getPercentiles() {
+		if(percentiles.size() > 4) {
+			log.warn("using more than 4 percentiles is heavily discouraged");
+		}
+		//Validate must be whole number between 50 and 100.
+		return percentiles.stream().mapToDouble(Double::parseDouble).map(e -> e / 100).toArray();
 	}
 
-	public void configureCache(Ehcache cache) {
-		new EhCache2Metrics(cache, root.tags).bindTo(registry);
+	@Override
+	public void destroy() throws Exception {
+		Search search = Search.in(registry).tag("configuration", applicationContext.getId());
+		search.counters().parallelStream().forEach(registry::remove);
+		search.gauges().parallelStream().forEach(registry::remove);
+		search.summaries().parallelStream().forEach(registry::remove);
```

Comment:
The usage of a `parallelStream` is only helping if the amount of data is really big, because it has overhead to join all results back together. Are you sure this is helping here? Do we have that many counters etc?

Reply:
No, and it's completely irrelevant in which order it removes meters 😶‍🌫️